libraries/devsec_ssh.rb

```diff
@@ -0,0 +1,220 @@
+
+      def respond_to_missing?(method, *)
+        type, call = parse_method(method)
+        type && call && CRYPTO.key?(call) && SSH_TYPES.include?(type)
+      end
+
+      # Parse the given method in form get_type_call, e.g. get_client_kexs
+      # Return the array with symbols [type, call]
+      def parse_method(method)
+        method.to_s.match(/^get_([[:alnum:]]*)_([[:alnum:]]*)$/) { |m| [m[1].to_sym, m[2].to_sym] }
+      end
+
+      # Find the ssh version, matching to the next small
+      # version in versions array
+      def find_ssh_version(version, versions)
+        found_ssh_version = nil
+        versions.map do |v|
+          next unless v.is_a?(Float) # skip everything what does not look like a version
+          found_ssh_version = v if version >= v
+        end
+
+        raise "Unsupported ssh version #{version}" unless found_ssh_version
+
+        found_ssh_version
+      end
+
+      { client: 'sshclient',
+        server: 'sshserver' }.each do |k, v|
+        define_method("get_ssh_#{k}_version") do |node|
+          get_ssh_version(node, node[v]['package'])
+        end
+      end
+
+      def get_ssh_version(node, package)
+        version = node['packages'][package]['version']
+        # on debian we get the epoch in front of version number: 1:7.2p2-4ubuntu2.1
+        version = version.split(':')[1] if node['platform_family'] == 'debian'
+        Chef::Log.debug("Detected openssh version #{version} for package #{package}")
+        version.to_f
+      rescue NoMethodError
+        # Detection via installed package failed, lets guess the installed version
+        version = guess_ssh_version(node)
+        Chef::Log.debug("Failed to get the openssh version from installed package. Guessed version is #{version}")
+        version
+      end
+
+      # Guess the version of ssh via OS matrix
+      def guess_ssh_version(node)
+        family = node['platform_family']
+        platform = node['platform']
+        version = node['platform_version'].to_f
+
+        case family
+        when 'debian'
+          case platform
+          when 'ubuntu'
+            return 6.6 if version >= 14.04
+          when 'debian'
+            return 6.6 if version >= 8
+            return 5.3 if version <= 6
+          end
+        when 'rhel'
+          return 6.6 if version >= 7
+          return 5.3 if version >= 6
+        end
+        Chef::Log.info("Unknown platform #{node['platform']} with version #{node['platform_version']} and family #{node['platform_family']}. Assuming ssh version #{FALLBACK_SSH_VERSION}")
+        FALLBACK_SSH_VERSION
+      end
+    end
+  end
+end
+
+class Chef
+  class Recipe
+    # This is a transparent connector from Chef Recipe
+    # to the implementation claas
+    #
+    # Functions from connected class can be used in the Chef recipes like
+    # DevSec.get_ciphers(node)
+    #
+    class DevSec
+      class << self
+        KLASS ||= ::DevSec::Ssh
+
+        def method_missing(method, *args)
```

Comment:
We can also get rid of this proxy connector and integrate the entire thing as `Chef::Recipe::DevSec::SSH`?